doc/bgp_loading_optimization/bgp-loading-optimization-hld.md

```diff
@@ -0,0 +1,421 @@
+The main thread of `zebra` not only needs to send routes to `fpmsyncd`, but also needs to process the returned results of the child thread which indicate whether data are successfully delivered to `kernel`. Hence when `zebra` is busy dealing with the `kernel` side, the performance of talking to `fpmsyncd` would be affected.
+
+
+<br>
+
+<figure align=center>
+    <img src="images/zebra.jpg" width="60%" height=auto>
+    <figcaption>Figure 6. Zebra flame graph<figcaption>
+</figure>  
+      
+### Synchronous sairedis API usage
+
+The interaction between `orchagent` and `syncd` is using synchronous `sairedis` API.
+Once `orchagent` `doTask` writes data to ASIC_DB, it waits for response from `syncd`. And since there is only single thread in `orchagent` it cannot process other routing messages until the response is received and processed.
+
+<figure align=center>
+    <img src="images/sync-sairedis1.png" width="40%" height=20%>
+    <figcaption>Figure 5. Sync sairedis workflow<figcaption>
+</figure> 
+
+
+## Requirements
+
+- All modifications should maintain the time sequence of route loading
+- All modules should support the warm restart operations after modified
+- With the optimization of this HLD implemented, the end-to-end BGP loading performance should be improved at least by 95%
+- The new optimization codes would be turn off by default. It could be turned on via configuration
+
+
+## High-Level Proposal
+
+### Modification in orchagent/syncd to enable multi-threading
+Figure 6 below illustrates the high level architecture modification for `orchagent` and `syncd`, it compares the original architecture and the new pipeline architecture proposed by this HLD. The pipeline design changes the workflow of both `orchagent` and `syncd`, thus enabling them to employ multiple threads to do sub-tasks concurrently.
+
+Take `orchagent` for example, a single task of `orchagent` contains three sub-tasks `pops`, `addToSync` and `doTask`, and originally `orchagent` performs the three sub-tasks in serial. A new `pops` sub-task can only begin after the previous `doTask` is finished. The proposed design utilizes a separate thread to run `pops`, which decouples the `pops` sub-task from `addToSync` and `doTask`. As the figure shows, in the new pipeline architecture, a new `pops` sub-task begins immediately when it's ready, not having to wait for the previous `addToSync` and `doTask` to finish.
+
+<figure align=center>
+    <img src="images/pipeline-timeline.png">
+    <figcaption>Figure 7. Pipeline architecture compared with the original serial architecture<figcaption>
+</figure>  
+
+#### Ring buffer for low-cost thread coordination
+Since multiple threads are employed, we take a lock-free design by using a ring buffer as an asynchronous communication channel.
+
+#### Asynchronous sairedis API usage
+Asynchronous mode `sairedis` API is used and a list of context of response pending messages is maintained on `orchagent` to process the response when its received
```

Comment:
>Asynchronous mode [](http://example.com/codeflow?start=0&length=17)

My understanding is that the communication between orchagent and syncd is handled by single syncd thread. How does async mode help?